File: 01-Arrays/03-tow-pointers/02-sliding-window/03_longest_substring_without_repeating.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    string s = "abcabcbb";
    unordered_set<char> chars;
    int l = 0, maxLen = 0;

    for (int r = 0; r < s.size(); ++r) {
        while (chars.count(s[r])) {
            chars.erase(s[l]);     
            l++;
        }
        chars.insert(s[r]);
        maxLen = max(maxLen, r - l + 1);
    }

    cout << maxLen << endl;
    return 0;
}

File: cp-python/03_tow-pointers/02-sliding_window/03_longest_substring_without_repeating.py
import sys
input = sys.stdin.readline
output = sys.stdout.write

s = "abcabcbb"
chars = set()
maxLen = 0
l = 0

for i,val in enumerate(s):
    while val in chars:
        chars.remove(s[l])
        l+=1
    chars.add(val)
    maxLen = max(maxLen, i - l + 1)


output(str(maxLen))
